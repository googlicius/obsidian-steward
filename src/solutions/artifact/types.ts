import { ContentReadingResult } from 'src/services/ContentReadingService';
import { IndexedDocument } from 'src/database/SearchDatabase';
import { ConditionResult } from 'src/solutions/search/searchEngineV3';
import { UpdateContentArgs } from 'src/solutions/commands/handlers/GenerateCommandHandler/zSchemas';

/**
 * Types of artifacts that can be stored for conversations
 */
export enum ArtifactType {
  SEARCH_RESULTS = 'search_results',
  MOVE_RESULTS = 'move_results',
  CALCULATION_RESULTS = 'calculation_results',
  CREATED_NOTES = 'created_notes',
  READ_CONTENT = 'read_content',
  CONTENT_UPDATE = 'content_update',
  GENERATED_CONTENT = 'generated_content',
  MEDIA_RESULTS = 'media_results',
  CONVERSATION_SUMMARY = 'conversation_summary',
  EXTRACTION_RESULT = 'extraction-result',
}

/**
 * Base interface for all artifacts
 */
export interface BaseArtifact {
  artifactType: ArtifactType;
  createdAt?: number; // Timestamp when the artifact was created
  id?: string; // ID of the artifact (usually the message ID that created it)
}

/**
 * Search results artifact
 */
export interface SearchResultsArtifact extends BaseArtifact {
  artifactType: ArtifactType.SEARCH_RESULTS;
  originalResults: ConditionResult<IndexedDocument>[]; // The original, unpaginated results
}

/**
 * Created notes artifact
 */
export interface CreatedNotesArtifact extends BaseArtifact {
  artifactType: ArtifactType.CREATED_NOTES;
  paths: string[];
}

/**
 * Read content artifact
 */
export interface ReadContentArtifact extends BaseArtifact {
  artifactType: ArtifactType.READ_CONTENT;
  readingResult: ContentReadingResult;
}

/**
 * Content update artifact
 */
export interface ContentUpdateArtifact extends BaseArtifact {
  artifactType: ArtifactType.CONTENT_UPDATE;
  path: string;
  updateExtraction: UpdateContentArgs;
}

/**
 * Media results artifact
 */
export interface MediaResultsArtifact extends BaseArtifact {
  artifactType: ArtifactType.MEDIA_RESULTS;
  paths: string[]; // Paths to the media files
  mediaType?: 'audio' | 'image'; // Type of media
}

/**
 * Conversation summary artifact
 */
export interface ConversationSummaryArtifact extends BaseArtifact {
  artifactType: ArtifactType.CONVERSATION_SUMMARY;
  summary: string; // The generated summary text
}

/**
 * Extraction result artifact
 */
export interface ExtractionResultArtifact extends BaseArtifact {
  artifactType: ArtifactType.EXTRACTION_RESULT;
  content: {
    query: string;
    commands: {
      commandType: string;
      query: string;
    }[];
  };
}

export type Artifact =
  | SearchResultsArtifact
  | CreatedNotesArtifact
  | ReadContentArtifact
  | ContentUpdateArtifact
  | MediaResultsArtifact
  | ConversationSummaryArtifact
  | ExtractionResultArtifact;

export type ArtifactMap = {
  [ArtifactType.SEARCH_RESULTS]: SearchResultsArtifact;
  [ArtifactType.CREATED_NOTES]: CreatedNotesArtifact;
  [ArtifactType.READ_CONTENT]: ReadContentArtifact;
  [ArtifactType.CONTENT_UPDATE]: ContentUpdateArtifact;
  [ArtifactType.MEDIA_RESULTS]: MediaResultsArtifact;
  [ArtifactType.CONVERSATION_SUMMARY]: ConversationSummaryArtifact;
  [ArtifactType.EXTRACTION_RESULT]: ExtractionResultArtifact;
};

/**
 * Interface for artifact serializers
 */
export interface ArtifactSerializer {
  /**
   * Serialize an artifact to a string wrapped in stw-artifact block
   * @param artifact The artifact to serialize
   * @returns The serialized artifact as a string wrapped in stw-artifact block
   */
  serialize(artifact: Artifact): string;

  /**
   * Deserialize a string to an artifact
   * @param data The string to deserialize (possibly containing a stw-artifact block)
   * @returns The deserialized artifact or a Promise that resolves to the artifact
   */
  deserialize(data: string): Artifact | Promise<Artifact>;
}
