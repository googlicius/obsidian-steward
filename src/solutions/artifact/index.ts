export * from './types';
export * from './ArtifactManagerV2';
