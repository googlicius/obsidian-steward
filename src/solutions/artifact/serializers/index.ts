export * from './JsonSerializer';
export * from './SearchResultSerializer';
export * from './CompositeSerializer';
