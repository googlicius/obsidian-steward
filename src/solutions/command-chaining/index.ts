export * from './ConversationContextManager';
export * from './CommandPipeline';
export * from './steps';
export * from './CommandFactory';
export * from './types';

// This file acts as a central export point for the command-chaining solution
