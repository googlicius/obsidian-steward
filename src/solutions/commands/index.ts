export * from './CommandProcessor';
export * from './CommandHandler';
export * from './handlers/SearchCommandHandler';
export * from './handlers/MoreCommandHandler';
export * from './handlers/MoveCommandHandler';
export * from './handlers/ConfirmCommandHandler';
