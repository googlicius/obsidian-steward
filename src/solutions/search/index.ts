export * from './documentStore';
export * from './tokenizer';
export * from './indexer';
export * from './scoring';
export * from './searchService';
export * from './stopwords';
