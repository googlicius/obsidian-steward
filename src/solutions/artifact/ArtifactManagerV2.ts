import { TFile } from 'obsidian';
import { logger } from 'src/utils/logger';
import type StewardPlugin from 'src/main';
import {
  Artifact,
  ArtifactMap,
  ArtifactSerializer,
  ArtifactType,
} from 'src/solutions/artifact/types';
import { JsonArtifactSerializer, SearchResultSerializer, CompositeSerializer } from './serializers';

/**
 * Manager for storing and retrieving artifacts in conversation notes
 */
export class ArtifactManagerV2 {
  private static instance: ArtifactManagerV2;
  private static withTitleInstances: Map<string, ArtifactManagerV2> = new Map();
  // Flag to track if the modify event listener has been registered
  private static modifyListenerRegistered = false;
  private conversationTitle = '';
  private serializers: Map<ArtifactType, ArtifactSerializer> = new Map();

  // In-memory cache of artifacts for the current conversation
  private artifactsCache: Artifact[] | null = null;

  get documentStore() {
    return this.plugin.searchService.documentStore;
  }

  constructor(
    private plugin: StewardPlugin,
    serializers?: Map<ArtifactType, ArtifactSerializer>
  ) {
    // Register default serializers
    if (!serializers) {
      for (const type of Object.values(ArtifactType)) {
        if (type === ArtifactType.SEARCH_RESULTS) {
          const searchResultSerializer = new SearchResultSerializer(this.documentStore);
          const jsonSerializer = new JsonArtifactSerializer(type);

          this.registerSerializer(
            type,
            new CompositeSerializer(searchResultSerializer, jsonSerializer)
          );
        } else {
          this.registerSerializer(type, new JsonArtifactSerializer(type));
        }
      }
    } else {
      this.serializers = serializers;
    }
  }

  /**
   * Register the modify event listener for all instances
   */
  private static registerModifyListener(plugin: StewardPlugin): void {
    if (ArtifactManagerV2.modifyListenerRegistered) {
      return;
    }

    plugin.registerEvent(
      plugin.app.vault.on('modify', async (file: TFile) => {
        const title = plugin.conversationRenderer.extractTitleFromPath(file.path);

        // Get the instance directly by title
        const manager = ArtifactManagerV2.withTitleInstances.get(title);

        // If we have a manager for this title and it has a cache, check if we need to refresh
        if (manager && manager.artifactsCache && manager.artifactsCache.length > 0) {
          const lastItem = manager.artifactsCache[manager.artifactsCache.length - 1];
          if (lastItem && lastItem.id) {
            const fileContent = await plugin.app.vault.cachedRead(file);
            if (!fileContent.includes(lastItem.id)) {
              logger.log(
                `Last item in cache for "${title}" does not exist in conversation note, refreshing cache`
              );
              manager.getAllArtifacts(true);
            }
          }
        }
      })
    );

    ArtifactManagerV2.modifyListenerRegistered = true;
    logger.log('Registered modify event listener for ArtifactManagerV2');
  }

  /**
   * Get the singleton instance
   */
  public static getInstance(plugin?: StewardPlugin): ArtifactManagerV2 {
    if (plugin) {
      ArtifactManagerV2.instance = new ArtifactManagerV2(plugin);
    }

    if (!ArtifactManagerV2.instance) {
      throw new Error('ArtifactManagerV2 is not initialized');
    }

    return ArtifactManagerV2.instance;
  }

  /**
   * Create a new instance or get the existing instance with the specified conversation title
   */
  public withTitle(conversationTitle: string): ArtifactManagerV2 {
    // Return existing instance if available
    if (ArtifactManagerV2.withTitleInstances.has(conversationTitle)) {
      return ArtifactManagerV2.withTitleInstances.get(conversationTitle) as ArtifactManagerV2;
    }

    // Create a new instance
    const manager = new ArtifactManagerV2(this.plugin, this.serializers);
    manager.conversationTitle = conversationTitle;
    manager.artifactsCache = null;

    // Register the instance in the static registry
    ArtifactManagerV2.withTitleInstances.set(conversationTitle, manager);

    // Ensure the modify event listener is registered
    ArtifactManagerV2.registerModifyListener(this.plugin);

    return manager;
  }

  /**
   * Register a custom serializer for an artifact type
   */
  public registerSerializer(type: ArtifactType, serializer: ArtifactSerializer): void {
    this.serializers.set(type, serializer);
  }

  /**
   * Get the serializer for a specific artifact type
   */
  private getSerializer(type: ArtifactType): ArtifactSerializer {
    const serializer = this.serializers.get(type);
    if (!serializer) {
      throw new Error(`No serializer registered for artifact type: ${type}`);
    }
    return serializer;
  }

  /**
   * Get the path to the conversation note
   */
  private getNotePath(): string {
    if (!this.conversationTitle) {
      throw new Error('Conversation title is not set. Use withTitle() first.');
    }
    return `${this.plugin.settings.stewardFolder}/Conversations/${this.conversationTitle}.md`;
  }

  /**
   * Get the file for the conversation note
   */
  private async getFile(): Promise<TFile> {
    const notePath = this.getNotePath();
    const file = this.plugin.app.vault.getFileByPath(notePath);

    if (!file) {
      throw new Error(`Note not found: ${notePath}`);
    }

    return file;
  }

  /**
   * Store an artifact in a conversation note
   */
  public async storeArtifact(params: {
    text?: string;
    artifact: Artifact;
  }): Promise<string | undefined> {
    try {
      if (!this.conversationTitle) {
        throw new Error('Conversation title is not set. Use withTitle() first.');
      }

      const file = await this.getFile();

      // Get the serializer for this artifact type
      const serializer = this.getSerializer(params.artifact.artifactType);

      // Build message metadata
      const { messageId, comment } = await this.plugin.conversationRenderer.buildMessageMetadata(
        this.conversationTitle,
        {
          type: 'artifact',
          role: 'assistant',
          artifactType: params.artifact.artifactType,
          includeHistory: false,
        }
      );

      if (!params.artifact.id) {
        params.artifact.id = messageId;
      }

      // Process the file content
      await this.plugin.app.vault.process(file, currentContent => {
        // Remove any generating indicator
        currentContent = this.plugin.conversationRenderer.removeGeneratingIndicator(currentContent);

        let contentToAdd = params.text ? `${params.text}\n` : '';

        contentToAdd += serializer.serialize(params.artifact);

        // Return the updated content
        return `${currentContent}\n\n${comment}\n${contentToAdd}`;
      });

      // Add to the in-memory cache
      await this.addToCache(params.artifact);

      logger.log('Stored artifact in conversation note', params.artifact);
      return messageId;
    } catch (error) {
      logger.error('Error storing artifact:', error);
      return undefined;
    }
  }

  /**
   * Add an artifact to the in-memory cache
   */
  private async addToCache(artifact: Artifact): Promise<void> {
    if (!this.artifactsCache) {
      this.artifactsCache = await this.getAllArtifacts(true);
    }
    this.artifactsCache.push(artifact);
  }

  /**
   * Extract all artifacts from a conversation note
   * @param refresh Force a refresh of the cache
   */
  public async getAllArtifacts(refresh = false): Promise<Artifact[]> {
    try {
      if (!this.conversationTitle) {
        throw new Error('Conversation title is not set. Use withTitle() first.');
      }

      // Check if we have a cached version and refresh is not requested
      if (!refresh && this.artifactsCache !== null) {
        return this.artifactsCache;
      }

      // Get all messages from the conversation
      const messages = await this.plugin.conversationRenderer.extractAllConversationMessages(
        this.conversationTitle
      );

      // Filter messages with type 'artifact'
      const artifactMessages = messages.filter(message => message.type === 'artifact');

      if (artifactMessages.length === 0) {
        this.artifactsCache = null;
        return [];
      }

      const artifacts: Artifact[] = [];

      for (const message of artifactMessages) {
        if (!message.artifactType) {
          continue;
        }
        try {
          const serializer = this.getSerializer(message.artifactType as ArtifactType);
          const result = await serializer.deserialize(message.content);
          artifacts.push(result);
        } catch (error) {
          logger.error(`Error deserializing artifact of type ${message.artifactType}:`, error);
        }
      }

      // Cache the results
      this.artifactsCache = artifacts;

      return artifacts;
    } catch (error) {
      logger.error('Error getting artifacts:', error);
      return [];
    }
  }

  /**
   * Get an artifact by its ID
   */
  public async getArtifactById(artifactId: string): Promise<Artifact | undefined> {
    // Get all artifacts from cache or load them if not available
    const artifacts = await this.getAllArtifacts();
    return artifacts.find(artifact => artifact.id === artifactId);
  }

  /**
   * Get the most recent artifact of a specific type
   */
  public async getMostRecentArtifactByType<T extends keyof ArtifactMap>(
    type: T
  ): Promise<ArtifactMap[T] | undefined> {
    // Get all artifacts from cache or load them if not available
    const artifacts = await this.getAllArtifacts();

    // Filter artifacts by type
    const typeArtifacts = artifacts.filter(artifact => artifact.artifactType === type);

    if (typeArtifacts.length === 0) {
      return undefined;
    }

    // Return the last artifact (most recently added)
    return typeArtifacts[typeArtifacts.length - 1] as ArtifactMap[T];
  }

  /**
   * Get the most recent artifact of specified types
   */
  public async getMostRecentArtifactOfTypes(types: ArtifactType[]): Promise<Artifact | undefined> {
    const artifacts = await this.getAllArtifacts();

    // Filter artifacts by the specified types
    const filteredArtifacts = artifacts.filter(artifact =>
      types.includes(artifact.artifactType as ArtifactType)
    );

    if (filteredArtifacts.length === 0) {
      return undefined;
    }

    return filteredArtifacts[filteredArtifacts.length - 1];
  }

  /**
   * Clear the in-memory cache
   */
  public clearCache(): void {
    this.artifactsCache = null;
    logger.log('Artifact cache cleared');
  }
}
