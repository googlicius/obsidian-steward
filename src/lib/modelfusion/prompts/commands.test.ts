import { ArtifactType } from 'src/solutions/artifact';
import {
  COMMAND_DEFINITIONS,
  CommandDefinition,
  artifactDependentExamples,
  formatCommandsForPrompt,
  formatCurrentArtifacts,
  formatQueryTemplatesForPrompt,
} from './commands';

describe('commands', () => {
  describe('formatCurrentArtifacts', () => {
    it('should return formatted current artifacts for prompt', () => {
      const formattedArtifacts = formatCurrentArtifacts([
        { artifactType: ArtifactType.SEARCH_RESULTS },
        { artifactType: ArtifactType.READ_CONTENT },
      ]);
      expect(formattedArtifacts).toMatchSnapshot();
    });
  });

  describe('formatCommandsForPrompt', () => {
    it('should return formatted commands for prompt', () => {
      const formattedCommands = formatCommandsForPrompt();
      expect(formattedCommands).toMatchSnapshot();
    });

    it('should return formatted commands for prompt with given command names', () => {
      const formattedCommands = formatCommandsForPrompt([
        'read',
        'generate',
        'update_from_artifact',
      ]);
      expect(formattedCommands).toMatchSnapshot();
    });

    it('should throw error if command not found', () => {
      expect(() => formatCommandsForPrompt(['not_found'])).toThrow('Command not_found not found');
    });
  });

  describe('formatQueryTemplatesForPrompt', () => {
    it('should return formatted query templates for prompt', () => {
      const formattedTemplates = formatQueryTemplatesForPrompt();
      expect(formattedTemplates).toMatchSnapshot();
    });

    it('should return formatted query templates for prompt with given command names', () => {
      const formattedTemplates = formatQueryTemplatesForPrompt([
        'read',
        'generate',
        'update_from_artifact',
      ]);
      expect(formattedTemplates).toMatchSnapshot();
    });

    it('should return empty YAML structure when no commands have templates', () => {
      const formattedTemplates = formatQueryTemplatesForPrompt(['close', 'confirm']);
      expect(formattedTemplates).toMatchSnapshot();
    });
  });

  describe('artifactDependentExamples', () => {
    it('should return examples for commands that depend on artifacts', () => {
      const commands = COMMAND_DEFINITIONS;
      const examples = artifactDependentExamples(commands);
      expect(examples).toMatchSnapshot();
    });

    it('should return empty string if there are no commands that depend on artifacts', () => {
      const commands = COMMAND_DEFINITIONS.filter(cmd => !cmd.artifactDesc);
      const examples = artifactDependentExamples(commands);
      expect(examples).toBe('');
    });

    it('should return empty string if there are no commands that create artifacts', () => {
      const commands = COMMAND_DEFINITIONS.filter(cmd => !cmd.artifactDesc);
      const examples = artifactDependentExamples(commands);
      expect(examples).toBe('');
    });

    it('should return only one example if there are only one command that creates artifacts', () => {
      const commands: CommandDefinition[] = [
        {
          commandType: 'read',
          description: 'Read a note',
          category: 'built-in',
          artifactDesc: 'The reading result',
        },
        {
          commandType: 'update_from_artifact',
          description: 'Update a note',
          category: 'built-in',
          artifactDesc: 'The updated content',
        },
      ];
      const examples = artifactDependentExamples(commands);
      expect(examples).toMatchSnapshot();
    });
  });
});
