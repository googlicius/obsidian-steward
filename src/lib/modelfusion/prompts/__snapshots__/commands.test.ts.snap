// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands artifactDependentExamples should return examples for commands that depend on artifacts 1`] = `"For example: if you include "search" and "move_from_artifact", the "move_from_artifact" command will use the artifact that will be created from the "search" command or "read" and "update_from_artifact", the "update_from_artifact" command will use the artifact that will be created from the "read" command."`;

exports[`commands artifactDependentExamples should return only one example if there are only one command that creates artifacts 1`] = `"For example: if you include "read" and "update_from_artifact", the "update_from_artifact" command will use the artifact that will be created from the "read" command."`;

exports[`commands formatCommandsForPrompt should return formatted commands for prompt 1`] = `
"- "search": Find files using the search engine to search files locally and store the result as an artifact
  Use when: Search for files (and doesn't mention existing search results)
- "close": Close the conversation or exit
  Use when: Close the conversation
- "image": Generate an image
  Use when: Generate an image
- "audio" (aliases: speak): Generate audio from text
  Use when: Generate audio
- "create": Create a new note with their own content
  Use when: Create a new note
- "move_from_artifact" (aliases: move): Move notes from the artifact to a destination
  Use when: Move notes from the artifact
- "copy_from_artifact" (aliases: copy): Copy notes from the artifact to a destination
  Use when: Copy notes from the artifact
- "update_from_artifact" (aliases: update): Update note(s) from the artifact
  Use when: Update notes from the artifact
- "delete_from_artifact" (aliases: delete): Delete note(s) from the artifact
  Use when: Delete notes from the artifact
- "build_search_index" (aliases: index, build-index, search-index): Build or rebuild the search index for all markdown files in the vault
  Use when: Build or rebuild the search index for all markdown files in the vault
- "generate": Generate content with the LLM help (either in a new note or in the conversation). You also can "generate" from the provided content in the user's query without reading the note. Example: "Help me update this list to the numbered list:\\n- Item 1\\n- Item 2" -> ["generate"]. The list is already in the query.
  Use when: Ask or generate content with your help
- "read": Read content from the current note or specific position: "above", "below". Use this when you don't know the content and need to retrieve it before proceeding
  Use when: Read or Find content based on a specific pattern in their current note"
`;

exports[`commands formatCommandsForPrompt should return formatted commands for prompt with given command names 1`] = `
"- "search": Find files using the search engine to search files locally and store the result as an artifact
  Use when: Search for files (and doesn't mention existing search results)
- "read": Read content from the current note or specific position: "above", "below". Use this when you don't know the content and need to retrieve it before proceeding
  Use when: Read or Find content based on a specific pattern in their current note"
`;

exports[`commands formatCurrentArtifacts should return formatted current artifacts for prompt 1`] = `
"Current artifacts in the conversation can be used by artifact-dependent commands:
- search_results
- read_content"
`;
